File: app/src/main/java/com/fstyle/structure_android/data/source/UserDataSource.java
package com.fstyle.structure_android.data.source;

import android.support.annotation.NonNull;
import com.fstyle.structure_android.data.model.User;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import java.util.List;

/**
 * Created by le.quang.dao on 10/03/2017.
 */

public interface UserDataSource {

    Flowable<List<User>> getUsers();

    /**
     * LocalData For User
     */
    interface LocalDataSource extends UserDataSource {

        Completable insertUser(@NonNull User user);

        Completable updateUser(@NonNull User user);

        Completable deleteUser(@NonNull User user);

        Completable insertOrUpdateUser(@NonNull User user);

        Flowable<User> getUser(String userLogin);

        Flowable<List<User>> searchUsers(String userName);

        Completable deleteAllUsers();
    }

    /**
     * RemoteData For User
     */
    interface RemoteDataSource extends UserDataSource {
    }
}

File: app/src/main/java/com/fstyle/structure_android/data/source/UserDataSource.kt
package com.fstyle.structure_android.data.source

import com.fstyle.structure_android.data.model.User
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single

/**
 * Created by le.quang.dao on 10/03/2017.
 */

interface UserDataSource {
  /**
   * LocalData For User
   */
  interface LocalDataSource : BaseLocalDataSource {

    fun insertUser(user: User): Completable

    fun updateUser(user: User): Completable

    fun deleteUser(user: User): Completable

    fun insertOrUpdateUser(user: User): Completable

    val allUser: Flowable<List<User>>

    fun getUserByUserLogin(userLogin: String): Single<User>
  }

  /**
   * RemoteData For User
   */
  interface RemoteDataSource {
    fun searchUsers(keyWord: String, limit: Int): Single<List<User>>
  }
}
